monkey/agent_plugins/credentials_collectors/chrome/src/windows_credentials_database_processor.py

```diff
@@ -1,16 +1,131 @@
+    with temporary_file() as temporary_database_path:
+        # copy database before querying it to bypass lock errors
+        shutil.copy(database_path, temporary_database_path)
+
+        yield from _extract_credentials(temporary_database_path)
+
+
+@contextmanager
+def temporary_file() -> Iterator[Path]:
+    file, path = tempfile.mkstemp()
+    os.close(file)
+    try:
+        yield Path(path)
+    finally:
+        os.remove(path)
+
+
+def _extract_credentials(temporary_database_path: Path) -> Iterator[ExtractedCredentialPair]:
+    try:
+        conn = sqlite3.connect(temporary_database_path)
+        for user, password in conn.execute(DATABASE_QUERY):
+            yield user, password
+    except Exception as err:
+        logger.error(f"Encountered an exception while connecting to database: {err}")
+    finally:
+        conn.close()
+
+
+DatabaseReader = Callable[[Path], Iterator[ExtractedCredentialPair]]
 
 
 class WindowsCredentialsDatabaseProcessor:
-    def __init__(self):
-        pass
+    def __init__(self, database_reader: DatabaseReader = get_logins_from_database):
+        self._read_logins_from_database = database_reader
 
     def __call__(
         self, interrupt: Event, database_paths: Collection[BrowserCredentialsDatabasePath]
     ) -> Collection[Credentials]:
-        return []
+        credentials = []
+
+        for item in interruptible_iter(database_paths, interrupt):
+            for user, password in self._read_logins_from_database(item.database_file_path):
+                decrypted_password = self._decrypt_password(password, item.master_key)
+                if user or decrypted_password:
+                    credentials.append((user, decrypted_password))
+        return [
+            Credentials(
+                identity=self._get_identity(user),
+                secret=self._get_password(password),
+            )
+            for user, password in set(credentials)
+        ]
+
+    @staticmethod
+    def _get_identity(user: str):
+        try:
+            return EmailAddress(email_address=user)
+        except ValueError:
+            return Username(username=user)
+
+    @staticmethod
+    def _get_password(password: Optional[str]) -> Optional[Password]:
+        if password is None:
+            return None
+        return Password(password=password)
+
+    def _decrypt_password(
+        self, encrypted_password: bytes, master_key: Optional[bytes]
+    ) -> Optional[str]:
+        decrypted_password = None
+        try:
+            if encrypted_password.startswith(b"v10"):  # chromium > v80
+                decrypted_password = self._decrypt_password_v80(encrypted_password, master_key)
+            else:
+                with suppress(Exception):
+                    password_bytes = win32crypt_unprotect_data(encrypted_password)
+                    if isinstance(password_bytes, bytes):
+                        decrypted_password = password_bytes.decode("utf-8")
+        except Exception as err:
```

Comment:
What exceptions could be raised?

Reply:
Updated to suppress those potential exceptions